### DISCUSSION TITLE
Support of encrypted OPC UA security policies

### DISCUSSION THREAD

**Discussion by vogler75** [apache/streampipes#1449] [General]
It seems that the OPC UA connection can only handle security policy "none". 
At least I am getting the message "No endpoint with security policy none".
I think there is a need to support also the other security policies.
I would say unencrypted communications are a "no-go".

**Comment by tenthe** [comment]
Hello @vogler75,
thank you for starting the discussion.
We currently support two levels of security for the OPC UA Adapter. None & Sign (username & password).
Certificates are not supported yet, but that would be a great addition to the OPC UA integration.

You can find the code for the adapter here [1]. 
We are currently using Milo for the connection, but we already had a brief discussion with the guys from PLC4X if we should change the library to PLC4X. Therefore, we should make sure that it supports all the features we require.

If you have experience in this area, we always welcome feedback to improve the implementation.

[1] https://github.com/apache/streampipes/blob/dev/streampipes-extensions/streampipes-connect-adapters-iiot/src/main/java/org/apache/streampipes/connect/iiot/adapters/opcua/OpcUaAdapter.java

**Reply by vogler75** [reply]
Hi Philipp,
yes I have some experiences with Milo from my project https://github.com/vogler75/automation-gateway/ :-)
I will try to get into streampipes, but I cannot say if I will find enough time for it :-(
I would not remove Milo and use PLC4X instead - with all respect to PLC4X.

**Reply by vogler75** [reply]
I have not yet read the documentation.... 
I think I would need a "developer bootcamp session" for it. 
How to setup a development environment. 
So that I can compile code and test it.

**Reply by tenthe** [reply]
Hello @vogler75,
your project looks very cool!
We have a development guide for working on the extension services, like OPC UA integration, see [1].
Please let me know if you have any issues or questions regarding the setup.

[1] https://cwiki.apache.org/confluence/display/STREAMPIPES/Extensions
